test-fixtures/combined_heap_limits.wat

```diff
@@ -0,0 +1,588 @@
+
+      ;; wheee! we get to add a header.
+      ;; first step in adding the header: figure out the name of the
+      ;; field. this involves using our little int2str helper function,
+      ;; which takes the value and a memory offset and returns the offset
+      ;; once it's done writing. it will *not* write the terminal null, so
+      ;; we'll need to do that.
+      (local.get $i)
+      (global.get $response_header_number_start_offset)
+      (call $int2str)
+      (i32.const 0)
+      (i32.store8)
+
+      ;; OK, our header string is set up, and we're just using a constant
+      ;; body, so we should be good to go.
+      (global.get $response_handle_buffer)
+      (i32.load)
+      (global.get $response_header_prefix)
+      (global.get $response_header_prefix) ;; \ These compute the length of the header, and
+      (call $strlen)                       ;; / put it on the stack
+      (global.get $lorem_ipsum)
+      (global.get $lorem_ipsum)            ;; \ these compute the length of the value, and
+      (call $strlen)                       ;; / put it on the stack
+      (call $response_add_header)
+      (call $maybe_error_die)
+
+      ;; OK, we added some data. subtract one from our counter and
+      ;; go again.
+      (local.get $i)
+      (i32.const 1)
+      (i32.sub)
+      (local.set $i)
+      (br $header_adding_loop))
+  
+    unreachable)
+
+
+  ;; add a body of the given size. we don't need to be too precious about
+  ;; this; HTTP bodies can be pretty much anything. so we're going to
+  ;; just write out 1k chunks from the start of our memory, and ignore
+  ;; the fact that this leaks internal state everywhere.
+  (func $add_body (param $size_in_kb i32)
+    (local $bytes_left_to_write i32)
+
+    ;; compute how many bytes to write. we do this because the spec
+    ;; allows partial writes, and I don't want to deal with writing
+    ;; an inner loop.
+    (local.get $size_in_kb)
+    (i32.const 1024)
+    (i32.mul)
+    (local.set $bytes_left_to_write)
+
+    ;; first thing's first: we need to create the body buffer inside
+    ;; the runtime
+    (global.get $response_body_buffer)
+    (call $response_new_body)
+    (call $maybe_error_die)
+
+    ;; now we're going to actually write to the buffer.
+    (block $body_write_block
+      (loop $body_write_loop
+        ;; if we've got 0 bytes left to write, let's just stop
+        (local.get $bytes_left_to_write)
+        (i32.const 0)
+        (i32.eq)
+        (br_if $body_write_block) ;; remember, this cancels execution of the
+                                  ;; rest of the block
+
+        ;; OK, we'll just tack on 1k more data
+        (global.get $response_body_buffer)
+        (i32.load) ;; push the handle
+        (i32.const 0) ;; push the pointer for the buffer; the start of memory,
+                      ;; because we don't care
+        (i32.const 1024) ;; the length of the buffer
+        (i32.const 0) ;; I think this means add it to the end?
```

Comment:
yup

Reply:
I figured, but I'll just grump that the assumption that `witx` enumerations are numbered in order from zero is not super clearly stated anywhere I could find.